### ISSUE TITLE
Tracking issue: Add array API standard support to the main namespace

### ISSUE THREAD

**Issue by mtsokol** [numpy/numpy#25076]
See the [slides from the NumPy 2.0 developer meeting](https://github.com/numpy/archive/blob/main/2.0_developer_meeting/NumPy_2.0_devmeeting_array_API_adoption.pdf) in the [numpy archive](https://github.com/numpy/archive/tree/main/2.0_developer_meeting) for an initial overview. That content has to be expanded into a NEP.

### Infra/support/docs

- [x] a CI job that runs https://github.com/data-apis/array-api-tests/ on the main namespace (https://github.com/numpy/numpy/pull/25167)
- [x] reference guide documentation
- [x] ~a user guide documentation page~

### main namespace

- [x] `np.bool` (numpy/numpy#25080)
- [x] `np.astype` (gh-25079)
- [x] `unique_all`, `unique_counts`, `unique_inverse`, `unique_values` (numpy/numpy#25088)
- [x] `concat`, `permute_dims` (numpy/numpy#25086)
- [x]  Trigonometric functions' aliases (numpy/numpy#25086)
- [x] `bitwise_left_shift`, `bitwise_right_shift`, `bitwise_invert` (numpy/numpy#25086)
- [x] `copy` keyword argument for `np.asarray` (https://github.com/numpy/numpy/pull/25168)

### `linalg` namespace

- [x] cross (https://github.com/numpy/numpy/pull/25145)
- [x] matmul (https://github.com/numpy/numpy/pull/25086)
- [x] matrix_norm (https://github.com/numpy/numpy/pull/25155)
- [x] matrix_transpose (https://github.com/numpy/numpy/pull/25155)
- [x] outer (https://github.com/numpy/numpy/pull/25101)
- [x] svdvals (gh-24940)
- [x] tensordot (https://github.com/numpy/numpy/pull/25086)
- [x] vecdot (https://github.com/numpy/numpy/pull/25155)
- [x] vector_norm (https://github.com/numpy/numpy/pull/25155)

### `ndarray` methods and attributes

- [x] `__array_namespace__` method (https://github.com/numpy/numpy/pull/25167)
- [x] `to_device` and `device` members (https://github.com/numpy/numpy/pull/25233)

### miscellaneous

- [x] `correction` argument for `var` and `std` (https://github.com/numpy/numpy/pull/25169)
- [x] `rtol` parameter for `pinv` and `upper` parameter for `cholesky` (https://github.com/numpy/numpy/pull/25388)

**Comment by asmeurer** [comment]
This mailing list thread goes into details for some of the "easy and useful" changes https://mail.python.org/archives/list/numpy-discussion@python.org/thread/TTZEUKXUICDHGTCX5EMR6DQTYOSDGRV7/#YKBWQ2AP76WYWAP6GFRYMPHZCKTC43KM

**Comment by rgommers** [comment]
> `__array_namespace__` definitions can differ from the normal definitions without making the transition story for libraries really harder

It looks to me like that would require adding a whole new hidden namespace, rather then a simple `return numpy`, potentially confusing, a lot of extra work, and the potential for extra bugs. Plus it would mean that the main namespace would actually _not_ be compatible, which has other side effects. So that does not sound all that appealing. I'll add it to the NEP draft (ETA within a week; hard to keep up with Mateusz:)) as a design alternative though.

**Comment by mtsokol** [comment]
Hi! I'm finalizing the last `np.linalg` Array API PR and I realized that although `vecdot` (vector dot) is available only in `np.array_api`:
```python
# Note: vecdot is not in NumPy
def vecdot(x1: Array, x2: Array, /, *, axis: int = -1) -> Array:
```
there's `np.vdot` that also claims to be vector dot, but with completely different semantics (it ravels input to 1-D arrays and doesn't have `axis` keyword and it's `Complex-conjugating dot product`). I find it a bit confusing to have `np.vdot`, `np.vector_dot` and `np.linalg.vector_dot`.
What do you think about renaming `vdot` to something else, like `conj_vector_dot`?

**Comment by rgommers** [comment]
`vdot` is a bit unfortunate indeed, but on the other hand we already have a lot of similar functions so adding yet another name perhaps does not help. The four array API functions should be orthogonal and complete, so I'd suggest applying the new `.. legacy::` admonition to `vdot` and de-emphasizing it when we reorganize the docs.

**Comment by rgommers** [comment]
We discussed this topic in the community meeting two days ago. Quick summary:

1. Let's get `np.bool` and the new CI job with xfails merged now,
2. For the other PRs, let's go ahead with the ones that are an improvement either way (independent of the array API standard)
3. For the ones that we would not do without the array API standard exists (e.g., adding `device`), those are blocked on writing/reviewing the NEP.
4. @seberg pointed out that it would be helpful to give a thumbs up on PRs from (2) that look fine, with 1-2 thumbs up emojis, it's easier to move ahead and merge them.

**Comment by lucascolley** [comment]
also, `numpy.fft.{fftfreq, rfftfreq}` need the `device` parameter.

**Comment by jakevdp** [comment]
Another TODO to add here: `np.sort()` and `np.argsort()` need a `descending` argument: https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.sort.html

**Comment by rgommers** [comment]
> I'm interested in what choice NumPy will make regarding the treatment of `NaN` entries when sorted in descending order, as this is not specified by the array API, but we'd like to make `jax.numpy` follow the `numpy` implementation in this respect.

Good question. Whether `nan`s go at the beginning or the end for `sort` output is undefined on purpose, and in numpy they go at the end. For consistency, I'd expect for `descending=False` the whole array to be reversed, so then `nan`s go at the beginning:
```python
>>> x = np.asarray([1.0, 2.0, np.nan, 3.0, 4.0, np.nan, 5.0])
>>> np.sort(x)
array([ 1.,  2.,  3.,  4.,  5., nan, nan])
>>> np.sort(x)[::-1]
array([nan, nan,  5.,  4.,  3.,  2.,  1.])
```

**Comment by jakevdp** [comment]
Great, thanks! That's exactly how I ended up implementing it here: https://github.com/google/jax/pull/19201

**Comment by rgommers** [comment]
This is all done, so let's close this tracking issue. Thanks all!

**Comment by mdhaber** [comment]
@rgommers Re: https://github.com/numpy/numpy/issues/25076#issuecomment-1877618972, was closing this without `sort`/`argsort` having a `descending` keyword intentional? Crossref gh-14728.

**Comment by rgommers** [comment]
Yes, intentional to close this tracking issue as "declaring victory for 2.0.0". For several of the loose ends we decided that individual issues could be used - tracking issues shouldn't stay open indefinitely.

IIRC `descending` was investigated and determined that it was a lot of work (because of `nan` sorting, as noted in gh-14728) and hence we punted on that for 2.0. @mtsokol you may be able to add some extra color here (I thought there was discussion elsewhere aside from gh-14728, but can't remember where)?

**Comment by mtsokol** [comment]
Apart from that issue I had one attempt to implement it, here: https://github.com/mtsokol/numpy/pull/5 but it turned out to be too much to bite at once. I also found another attempt here: https://github.com/numpy/numpy/pull/14989.

**Comment by lucascolley** [comment]
shall we open a new issue to track it? Or are there other array API standard compatibility things tracked somewhere?

**Comment by rgommers** [comment]
> shall we open a new issue to track it? Or are there other array API standard compatibility things tracked somewhere?

Please use gh-14728 for this `descending` one, adding content as needed. 

I'm not in favor of a permanent array API compatibility issue if that's what you meant, it'd be better to use the existing [label](https://github.com/numpy/numpy/issues?q=is%3Aopen+is%3Aissue+label%3A%2240+-+array+API+standard%22) for it.
